src/Microsoft.OData.Core/Json/ODataJsonWriterUtils.cs

```diff
@@ -356,5 +430,183 @@ private static void ODataResourceValueToString(StringBuilder sb, ODataResourceVa
+        /// <summary>
+        /// Asynchronously writes an error message.
+        /// </summary>
+        /// <param name="jsonWriter">JSON writer.</param>
+        /// <param name="code">The code of the error.</param>
+        /// <param name="message">The message of the error.</param>
+        /// <param name="target">The target of the error.</param>
+        /// <param name="details">The details of the error.</param>
+        /// <param name="innerError">Inner error details that will be included in debug mode (if present).</param>
+        /// <param name="instanceAnnotations">Instance annotations for this error.</param>
+        /// <param name="writeInstanceAnnotationsDelegate">Delegate to write the instance annotations.</param>
+        /// <param name="maxInnerErrorDepth">The maximum number of nested inner errors to allow.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        private static async Task WriteErrorAsync(
+            IJsonWriterAsync jsonWriter,
+            string code,
+            string message,
+            string target,
+            IEnumerable<ODataErrorDetail> details,
+            ODataInnerError innerError,
+            IEnumerable<ODataInstanceAnnotation> instanceAnnotations,
+            Func<IEnumerable<ODataInstanceAnnotation>, Task> writeInstanceAnnotationsDelegate,
+            int maxInnerErrorDepth)
+        {
+            Debug.Assert(jsonWriter != null, "jsonWriter != null");
+            Debug.Assert(code != null, "code != null");
+            Debug.Assert(message != null, "message != null");
+            Debug.Assert(instanceAnnotations != null, "instanceAnnotations != null");
+
+            // "error": {
+            await jsonWriter.StartObjectScopeAsync().ConfigureAwait(false);
+            await jsonWriter.WriteNameAsync(JsonLightConstants.ODataErrorPropertyName).ConfigureAwait(false);
+            await jsonWriter.StartObjectScopeAsync().ConfigureAwait(false);
+
+            // "code": "<code>"
+            await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorCodeName).ConfigureAwait(false);
+            await jsonWriter.WriteValueAsync(code).ConfigureAwait(false);
+
+            // "message": "<message string>"
+            await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorMessageName).ConfigureAwait(false);
+            await jsonWriter.WriteValueAsync(message).ConfigureAwait(false);
+
+            // For example, "target": "query",
+            if (target != null)
+            {
+                await jsonWriter.WriteNameAsync(JsonConstants.ODataErrorTargetName).ConfigureAwait(false);
+                await jsonWriter.WriteValueAsync(target).ConfigureAwait(false);
+            }
+
+            // Such as, "details": [
+            //  {
+            //   "code": "301",
+            //   "target": "$search",
+            //   "message": "$search query option not supported"
+            //  }]
+            if (details?.Any() == true)

```

Comment:
>?.Any() == true [](start = 23, length = 15)

probably better to handle empty details in WriteErrorDetailsAsync, rather than incurring cost of the LINQ Any() method prior to calling.

Reply:
@mikepizzo I had thought that it'd be less expensive to call Any() method instead of calling await since that in turn would trigger creation of the state machine and the even more expensive GC clean up? What are your thoughts on that?

Reply:
Partially depends on how details (and instanceAnnotations) are constructed.  details are probably pretty static and easy to enumerate, but instanceAnnotations may be expensive to enumerate (i.e., if the enumerable is a filter over a broader set of instance annotations).

---
In reply to: [603521432](https://github.com/OData/odata.net/pull/2031#discussion_r603521432) [](ancestors = 603521432)